feat(lite): Tasks

ByScripts opened the pull request:
⚠️ Wait for #6620 before merge

![Capture_2023-02-27_11:20:16](https://user-images.githubusercontent.com/10992860/221537606-902bc194-3655-49cb-8409-2849ba3efd9c.png)

### Check list

> Check if done, if not relevant leave unchecked.

- [ ] PR reference the relevant issue (e.g. `Fixes #007` or `See xoa-support#42`)
- [ ] if UI changes, a screenshot has been added to the PR
- [ ] documentation updated
- `CHANGELOG.unreleased.md`:
  - [ ] enhancement/bug fix entry added
  - [ ] list of packages to release updated (`${name} v${new version}`)
- **I have tested added/updated features** (and impacted code)
  - [ ] unit tests (e.g. [`cron/parse.spec.js`](https://github.com/vatesfr/xen-orchestra/blob/b24400b21de1ebafa1099c56bac1de5c988d9202/%40xen-orchestra/cron/src/parse.spec.js))
  - [ ] if `xo-server` API changes, the corresponding test has been added to/updated on [`xo-server-test`](https://github.com/vatesfr/xen-orchestra/tree/master/packages/xo-server-test)
  - [ ] at least manual testing

### Process

1. create a PR as soon as possible
1. mark it as `WiP:` (Work in Progress) if not ready to be merged
1. when you want a review, add a reviewer (and only one)
1. if necessary, update your PR, and re- add a reviewer

From [_the Four Agreements_](https://en.wikipedia.org/wiki/Don_Miguel_Ruiz#The_Four_Agreements):

1. Be impeccable with your word.
1. Don't take anything personally.
1. Don't make assumptions.
1. Always do your best.

ByScripts committed:
feat(lite): RelativeTime component

ByScripts committed:
feat(lite): Tasks

MathieuRA commented on @xen-orchestra/lite/src/locales/en.json:
```diff
@@ -8,18 +8,15 @@
   "available-properties-for-advanced-filter": "Available properties for advanced filter:",
-  "backup": "Backup",
   "back-pool-dashboard": "Go back to your Pool dashboard",
+  "backup": "Backup",
```
I am not sure it's a good idea to reorder, it will create lot of conflict i guess.
If we keep this idea, you sould reorder `appearance` to be sync with `fr.json`

ByScripts replied:
I think ordering keys in precisely the right thing to do to avoid most conflicts :) 

Regarding the `appearance` key, I'm not sure to understand what you mean. The keys ordering seems to be exactly the same in [fr.json](https://raw.githubusercontent.com/vatesfr/xen-orchestra/af9437caa4914ffaed6af5fb629051aa6283d104/%40xen-orchestra/lite/src/locales/fr.json) and [en.json](https://raw.githubusercontent.com/vatesfr/xen-orchestra/af9437caa4914ffaed6af5fb629051aa6283d104/%40xen-orchestra/lite/src/locales/en.json).

MathieuRA commented on @xen-orchestra/lite/src/locales/en.json:
```diff
@@ -68,7 +65,17 @@
+  "task": {
+    "estimated-end": "Estimated end",
+    "name": "Name",
+    "object": "Object",
```
They will probably be used elsewhere, it would be better not to "link" them to task and make them "generic"

ByScripts replied:
It was my first idea. But since there is no "context information", I was afraid of a potential future misunderstanding so I thought it was preferable to encapsulate `name` and `object` to their respective group.

Maybe we should discuss this? And perhaps we should consider moving from JSON to JS, to be able to add some context comments.

```js
{
  ...
  name: "Name" // As in VM name, Host name...
  ...
}
```

ByScripts replied:
As discussed together, I did the change for `name` and `object` since there should not be any conflict of meaning of these specific terms.

MathieuRA commented on @xen-orchestra/lite/src/components/tasks/PendingTaskRow.vue:
```diff
@@ -0,0 +1,51 @@
+
+const props = defineProps<{
+  task: XenApiTask;
+}>();
```
Isn't it better to just ask for `task.uuid` instead of passing the full task?

ByScripts replied:
We are already looping on a list of tasks.

By passing only the uuid instead of the task, we will need to query the store to retrieve the same task, for each row.

ByScripts replied:
As discussed together, I'll do some performance checks later to ensure there is no perf issue by passing the task.

MathieuRA commented on @xen-orchestra/lite/src/components/tasks/TasksTable.vue:
```diff
@@ -0,0 +1,36 @@
+      v-for="task in finishedTasks"
+      :key="task.uuid"
+      :task="task"
+    />
```
IMO, since these two components are identical, we should mutualize them.
```js
  <TaskRow
      v-for="task in pendingTasks"
      :key="task.uuid"
      :task="task"
    />
  <TaskRow
    class="finished-tasks"
    v-for="task in finishedTasks"
    :key="task.uuid"
    :task="task"
  />
```

```css
.finished-tasks :deep(td) {
  opacity: 0.5;
}
```

ByScripts replied:
`PendingTaskRow` and `FinishedTaskRow` have more differences than just the opacity.

They don't compute the same data, and don't display the same cells content.

I my first attempt I had a single component for both, but it was sufficiently less readable (with a lot of conditions based on a prop) to separate them.

ByScripts replied:
I mutualized the components

MathieuRA commented on @xen-orchestra/lite/src/components/tasks/TasksTable.vue:
```diff
@@ -0,0 +1,36 @@
+}>();
+</script>
+
+<style lang="postcss" scoped></style>
```
In order to adapt to the FIGMA model:
```suggestion
<style lang="postcss" scoped>
:deep(td),
:deep(th) {
  border-right: none;
}
</style>
```

ByScripts replied:
There is indeed a difference between some tables in the Figma models 👍🏻 

Therefore, I think it would be preferable to add a prop to `UiTable` for vertical separator.

Since the majority of tables don't have the vertical separator, I'll make it an opt-in.

ByScripts replied:
I'll update UiTable on another PR to add this prop and to set the width to 100%.

MathieuRA commented on @xen-orchestra/lite/src/components/tasks/TasksTable.vue:
```diff
@@ -0,0 +1,36 @@
+      <th>{{ $t("task.object") }}</th>
+      <th>{{ $t("task.progress") }}</th>
+      <th>{{ $t("task.started") }}</th>
+      <th>{{ $t("task.estimated-end") }}</th>
```
Missing icons.
If it is voluntary, it would be necessary to add a comment on FIGMA to ask Clémence to remove them too.

MathieuRA reviewed (changes requested):
IMO the table should have a fixed width (100% of the available space to match the figma design)

ByScripts committed:
Mutualization of `FinishedTaskRow` and `PendingTaskRow` components

pdonias commented on @xen-orchestra/lite/src/components/tasks/TaskRow.vue:
```diff
@@ -0,0 +1,60 @@
+const estimatedEndAt = computed(
+  () =>
+    createdAt.value +
+    (new Date().getTime() - createdAt.value) / props.task.progress
```
Is `props.task.progress` always defined?

ByScripts replied:
I've never seen it undefined. Is there any way to be sure?

pdonias commented on @xen-orchestra/lite/src/views/pool/PoolTasksView.vue:
```diff
@@ -1 +1,69 @@
+
+const finishedTasks = computed(() => tasksHistory.value.slice(-10));
+
+useTitle(computed(() => `(${pendingTasks.value.length}) Tasks`));
```
This should be internationalized too.

pdonias commented on @xen-orchestra/lite/src/views/pool/PoolTasksView.vue:
```diff
@@ -1 +1,69 @@
+
+const tasksHistory = useArrayHistory(allTasks, 500, (task) => task.uuid);
+
+const finishedTasks = computed(() => tasksHistory.value.slice(-10));
```
I don't get this, can you explain?

ByScripts replied:
I made that to (manually) test pagination and forgot to remove.

`tasksHistory` would contains all finished tasks and `finishedTasks` would contains only the one to be displayed (the naming is really bad too, but was not intended to be published ^^).

For now, we'll keep only the history since pagination is not part of this card.

ByScripts committed:
Rewrite relative time

ByScripts committed:
Tasks page title i18n

ByScripts committed:
Remove unneeded code

ByScripts committed:
feat(lite/tasks)

- Added end date time to finished tasks
- Added callback to transform objects added to array history
- Fix bug on history value
- Allow parseDateTime to accept format from toISOString()

ByScripts commented:
I fixed conflicts.

@pdonias Can this be merged, or does it need further investigations?
